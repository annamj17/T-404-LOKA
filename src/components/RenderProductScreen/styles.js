import { StyleSheet } from 'react-native';

export default StyleSheet.create({
  screens: {
    flexDirection: 'column',
    height: '100%',
    width: '100%'
  },
  image: {
    marginTop: 30,
    height: '35%',
    width: '70%',
    alignSelf: 'center',
    borderRadius: 10
  },
  imageContainer: {
    marginTop: 'auto',
    height: '12%'
  },
  tagContainer: {
    flexDirection: 'row',
    height: 35,
    fontSize: 18,
    color: 'grey',
    fontWeight: 'bold'
  },
  product: {
    alignSelf: 'center',
    color: 'black',
    fontWeight: 'bold',
    fontSize: 18
  },
  tagItem: {
    marginLeft: 20,
    fontSize: 18,
    color: 'grey',
    fontWeight: 'bold',
    borderBottomColor: 'gray',
    borderBottomWidth: 1,
    width: 100,
    alignSelf: 'center'
  },
  scoreItem: {
    marginLeft: 10,
    fontSize: 16,
    color: 'grey',
    fontWeight: 'bold',
    borderBottomColor: 'gray',
    alignSelf: 'center'
  },
  priceItem: {
    marginTop: 15,
    marginLeft: 10,
    fontSize: 18,
    color: 'grey',
    fontWeight: 'bold',
    alignSelf: 'center'
  },
  priceTag: {
    marginTop: 15,
    alignSelf: 'center',
    color: 'black',
    fontWeight: 'bold',
    fontSize: 18
  },
  confirmContainer: {
    alignItems: 'center',
    justifyContent: 'center',
    alignSelf: 'center',
    backgroundColor: '#318F95',
    borderRadius: 10,
    marginTop: 20,
    height: 50,
    width: 190
  },
  confirmText: {
    color: 'white',
    fontWeight: 'bold',
    fontSize: 20
  },
  container: {
    backgroundColor: 'white',
    flex: 1
  },
  error: {
    color: 'red',
    textAlign: 'center',
    fontSize: 10
  }
});
