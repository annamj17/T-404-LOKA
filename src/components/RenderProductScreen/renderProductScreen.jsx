import React from 'react';
import { View, Text, Image, TextInput, TouchableOpacity, KeyboardAvoidingView } from 'react-native';
import { Card } from 'react-native-elements';
import styles from './styles';

import bottomImage from '../../../assets/Capture.png';
import { getAllTagsAfterEdit } from '../../services/editTagService';

class RenderProductScreen extends React.Component {
  constructor(props) {
    super(props);
    const { productData } = props;
    this.state = {
      productDetailsId: productData.productDetailsId,
      color: productData.color,
      colorScore: productData.colorScore,
      brand: productData.brand,
      brandScore: productData.brandScore,
      type: productData.type,
      typeScore: productData.typeScore,
      productPrice: productData.productPrice,
      error: ''
    };
  }

  // Handles onChangeText for input
  onInputHandler(name, value) {
    this.setState({ [name]: value });
  }

  // Handles if tags are edited or not and navigates to appropriate view
  onPress = async () => {
    const { navigation } = this.props;
    const { productData } = this.props;
    const originalData = productData;
    const {
      productDetailsId,
      color,
      brand,
      type,
      colorScore,
      brandScore,
      typeScore,
      productPrice
    } = this.state;
    const editedProductTags = {
      productDetailsId,
      color,
      brand,
      type,
      colorScore,
      brandScore,
      typeScore,
      productPrice
    };

    const id = productDetailsId;
    // If there is no editing
    if (
      editedProductTags.color === originalData.color &&
      editedProductTags.type === originalData.type &&
      editedProductTags.brand === originalData.brand
    ) {
      if (color && brand && type !== '') {
        const tags = editedProductTags;
        navigation.navigate('FinalView', { tags });
      }
    }
    // If there is editing
    if (
      editedProductTags.color !== originalData.color ||
      editedProductTags.type !== originalData.type ||
      editedProductTags.brand !== originalData.brand
    ) {
      if (color && brand && type !== '') {
        const tags = await getAllTagsAfterEdit(editedProductTags, id);
        navigation.navigate('FinalView', { tags });
      } else {
        this.errorValidation();
      }
    }
  };

  // Error validation if input field is empty
  errorValidation() {
    this.setState({
      error: 'All fields are required!'
    });
  }

  render() {
    const { photoData } = this.props;
    const { productData } = this.props;
    const { color, brand, type, productPrice, error } = this.state;
    const scoreColor = `${(productData.colorScore * 100).toFixed(2)}%`;
    const scoreType = `${(productData.typeScore * 100).toFixed(2)}%`;
    const scoreBrand = `${(productData.brandScore * 100).toFixed(2)}%`;

    return (
      <View style={styles.screens}>
        <KeyboardAvoidingView style={styles.container} behavior="padding">
          <Image source={{ uri: photoData }} style={styles.image} />
          <Card containerStyle={{ padding: 20 }}>
            <View style={styles.tagContainer}>
              <Text style={styles.product}>Color:</Text>
              <TextInput
                style={styles.tagItem}
                maxLength={20}
                name="color"
                defaultValue={color}
                clearButtonMode="while-editing"
                value={color}
                onChangeText={text => this.onInputHandler('color', text)}
              />
              <Text style={styles.scoreItem}>{`(${scoreColor})`}</Text>
            </View>
            <View style={styles.tagContainer}>
              <Text style={styles.product}>Type:</Text>
              <TextInput
                style={styles.tagItem}
                name="type"
                maxLength={20}
                defaultValue={type}
                clearButtonMode="while-editing"
                value={type}
                onChangeText={text => this.onInputHandler('type', text)}
              />
              <Text style={styles.scoreItem}>{`(${scoreType})`}</Text>
            </View>
            <View style={styles.tagContainer}>
              <Text style={styles.product}>Brand:</Text>
              <TextInput
                style={styles.tagItem}
                name="brand"
                maxLength={20}
                defaultValue={brand}
                clearButtonMode="while-editing"
                value={brand}
                onChangeText={text => this.onInputHandler('brand', text)}
              />
              <Text style={styles.scoreItem}>{`(${scoreBrand})`}</Text>
            </View>
            <Text style={styles.error}>{error}</Text>
            <View style={styles.tagContainer}>
              <Text style={styles.priceTag}>Full Condition Price:</Text>
              <TextInput
                style={styles.priceItem}
                name="fullConditionPrice"
                maxLength={20}
                editable={false}
                defaultValue={`${productPrice.fullConditionPrice}`}
                clearButtonMode="while-editing"
                value={`${productPrice.fullConditionPrice}£`}
              />
            </View>
            <View style={styles.tagContainer}>
              <Text style={styles.priceTag}>Resale Price:</Text>
              <TextInput
                style={styles.priceItem}
                name="resalePrice"
                maxLength={20}
                editable={false}
                defaultValue={`${productPrice.resalePrice}`}
                clearButtonMode="while-editing"
                value={`${productPrice.resalePrice}£`}
              />
            </View>
          </Card>
          <TouchableOpacity
            style={styles.confirmContainer}
            name="register"
            onPress={() => this.onPress()}
          >
            <Text style={styles.confirmText}>Register Item</Text>
          </TouchableOpacity>
        </KeyboardAvoidingView>
        <Image style={styles.imageContainer} source={bottomImage} resizeMode="stretch" />
      </View>
    );
  }
}
export default RenderProductScreen;
