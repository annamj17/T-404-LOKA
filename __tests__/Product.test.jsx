import React from 'react';
import { TextInput } from 'react-native';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

import RenderProductScreen from '../src/components/RenderProductScreen/renderProductScreen';

describe('Should render the Product page without crashing', () => {
  it('renders correctly', () => {
    const wrapper = renderer
      .create(
        <RenderProductScreen
          productData={{
            productDetailsId: '1',
            color: 'Black',
            colorScore: '99.98',
            brand: 'Nike',
            brandScore: '99.78',
            type: 'Sweatshirt',
            typeScore: '98.99',
            productPrice: {
              fullConditionPrice: '50',
              resalePrice: '25'
            }
          }}
        />
      )
      .toJSON();
    expect(wrapper).toMatchSnapshot();
  });
});

describe('Product tests', () => {
  let wrapper;
  const onNavigateSpy = jest.fn();
  const props = {
    navigation: {
      navigate: onNavigateSpy
    }
  };

  beforeEach(() => {
    wrapper = shallow(
      <RenderProductScreen
        productData={{
          productDetailsId: '1',
          color: 'Black',
          colorScore: '99.98',
          brand: 'Nike',
          brandScore: '99.78',
          type: 'Sweatshirt',
          typeScore: '98.99',
          productPrice: {
            fullConditionPrice: '50',
            resalePrice: '25'
          },
          error: ''
        }}
        {...props}
      />
    );
  });

  it('Should find 5 TextInputs', () => {
    expect(wrapper.find('TextInput')).toHaveLength(5);
  });

  it('Should find the TextInput for color', () => {
    expect(wrapper.find('TextInput[name="color"]')).toHaveLength(1);
  });

  it('Should test TextInput for color', () => {
    expect(shallow(<TextInput color="Black" />).prop('color')).toBe('Black');
  });

  it('Should find the TextInput for brand', () => {
    expect(wrapper.find('TextInput[name="brand"]')).toHaveLength(1);
  });

  it('Should test TextInput for brand', () => {
    expect(shallow(<TextInput brand="Adidas" />).prop('brand')).toBe('Adidas');
  });

  it('Should find the TextInput for type', () => {
    expect(wrapper.find('TextInput[name="type"]')).toHaveLength(1);
  });

  it('Should test TextInput for type', () => {
    expect(shallow(<TextInput type="Sweatshirt" />).prop('type')).toBe('Sweatshirt');
  });

  it('Should find the TextInput for fullConditionPrice', () => {
    expect(wrapper.find('TextInput[name="fullConditionPrice"]')).toHaveLength(1);
  });

  it('Should test TextInput for fullConditionPrice', () => {
    expect(shallow(<TextInput fullConditionPrice="50" />).prop('fullConditionPrice')).toBe('50');
  });

  it('Should find the TextInput for resalePrice', () => {
    expect(wrapper.find('TextInput[name="resalePrice"]')).toHaveLength(1);
  });

  it('Should test TextInput for resalePrice', () => {
    expect(shallow(<TextInput resalePrice="25" />).prop('resalePrice')).toBe('25');
  });

  it('Should find 1 TouchableOpacity', () => {
    expect(wrapper.find('TouchableOpacity')).toHaveLength(1);
  });

  it('Should handle onPressFunction', () => {
    const instance = wrapper.instance();
    jest.spyOn(instance, 'onPress');
    wrapper.find('TouchableOpacity[name="register"]').simulate('Press');
    expect(instance.onPress).toHaveBeenCalled();
  });

  it('Should handle onInputHandler for color', () => {
    const instance = wrapper.instance();
    jest.spyOn(instance, 'onInputHandler');
    wrapper.find('TextInput[name="color"]').simulate('changeText', 'blue');
    expect(instance.onInputHandler).toHaveBeenCalledWith('color', 'blue');
  });

  it('Should handle onInputHandler for type', () => {
    const instance = wrapper.instance();
    jest.spyOn(instance, 'onInputHandler');
    wrapper.find('TextInput[name="type"]').simulate('changeText', 'hoodie');
    expect(instance.onInputHandler).toHaveBeenCalledWith('type', 'hoodie');
  });

  it('Should handle onInputHandler for brand', () => {
    const instance = wrapper.instance();
    jest.spyOn(instance, 'onInputHandler');
    wrapper.find('TextInput[name="brand"]').simulate('changeText', 'adidas');
    expect(instance.onInputHandler).toHaveBeenCalledWith('brand', 'adidas');
  });

  it('Should handle onInputHandler', () => {
    const instance = wrapper.instance();
    jest.spyOn(instance, 'onInputHandler');
    wrapper.find('TextInput[name="brand"]').simulate('changeText', 'adidas');
    expect(instance.onInputHandler).toHaveBeenCalledWith('brand', 'adidas');
  });
});
