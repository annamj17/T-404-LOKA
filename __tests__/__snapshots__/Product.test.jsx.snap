// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render the Product page without crashing renders correctly 1`] = `
<View
  style={
    Object {
      "flexDirection": "column",
      "height": "100%",
      "width": "100%",
    }
  }
>
  <View
    onLayout={[Function]}
    style={
      Array [
        Object {
          "backgroundColor": "white",
          "flex": 1,
        },
        Object {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <Image
      source={
        Object {
          "uri": undefined,
        }
      }
      style={
        Object {
          "alignSelf": "center",
          "borderRadius": 10,
          "height": "35%",
          "marginTop": 30,
          "width": "70%",
        }
      }
    />
    <View
      style={
        Object {
          "backgroundColor": "white",
          "borderColor": "#e1e8ee",
          "borderWidth": 1,
          "margin": 15,
          "marginBottom": 0,
          "padding": 20,
          "shadowColor": "rgba(0,0,0, .2)",
          "shadowOffset": Object {
            "height": 0,
            "width": 0,
          },
          "shadowOpacity": 1,
          "shadowRadius": 1,
        }
      }
    >
      <View
        style={
          Object {
            "backgroundColor": "transparent",
          }
        }
      >
        <View
          style={
            Object {
              "color": "grey",
              "flexDirection": "row",
              "fontSize": 18,
              "fontWeight": "bold",
              "height": 35,
            }
          }
        >
          <Text
            style={
              Object {
                "alignSelf": "center",
                "color": "black",
                "fontSize": 18,
                "fontWeight": "bold",
              }
            }
          >
            Color:
          </Text>
          <TextInput
            allowFontScaling={true}
            clearButtonMode="while-editing"
            defaultValue="Black"
            maxLength={20}
            name="color"
            onChangeText={[Function]}
            rejectResponderTermination={true}
            style={
              Object {
                "alignSelf": "center",
                "borderBottomColor": "gray",
                "borderBottomWidth": 1,
                "color": "grey",
                "fontSize": 18,
                "fontWeight": "bold",
                "marginLeft": 20,
                "width": 100,
              }
            }
            underlineColorAndroid="transparent"
            value="Black"
          />
          <Text
            style={
              Object {
                "alignSelf": "center",
                "borderBottomColor": "gray",
                "color": "grey",
                "fontSize": 16,
                "fontWeight": "bold",
                "marginLeft": 10,
              }
            }
          >
            (9998.00%)
          </Text>
        </View>
        <View
          style={
            Object {
              "color": "grey",
              "flexDirection": "row",
              "fontSize": 18,
              "fontWeight": "bold",
              "height": 35,
            }
          }
        >
          <Text
            style={
              Object {
                "alignSelf": "center",
                "color": "black",
                "fontSize": 18,
                "fontWeight": "bold",
              }
            }
          >
            Type:
          </Text>
          <TextInput
            allowFontScaling={true}
            clearButtonMode="while-editing"
            defaultValue="Sweatshirt"
            maxLength={20}
            name="type"
            onChangeText={[Function]}
            rejectResponderTermination={true}
            style={
              Object {
                "alignSelf": "center",
                "borderBottomColor": "gray",
                "borderBottomWidth": 1,
                "color": "grey",
                "fontSize": 18,
                "fontWeight": "bold",
                "marginLeft": 20,
                "width": 100,
              }
            }
            underlineColorAndroid="transparent"
            value="Sweatshirt"
          />
          <Text
            style={
              Object {
                "alignSelf": "center",
                "borderBottomColor": "gray",
                "color": "grey",
                "fontSize": 16,
                "fontWeight": "bold",
                "marginLeft": 10,
              }
            }
          >
            (9899.00%)
          </Text>
        </View>
        <View
          style={
            Object {
              "color": "grey",
              "flexDirection": "row",
              "fontSize": 18,
              "fontWeight": "bold",
              "height": 35,
            }
          }
        >
          <Text
            style={
              Object {
                "alignSelf": "center",
                "color": "black",
                "fontSize": 18,
                "fontWeight": "bold",
              }
            }
          >
            Brand:
          </Text>
          <TextInput
            allowFontScaling={true}
            clearButtonMode="while-editing"
            defaultValue="Nike"
            maxLength={20}
            name="brand"
            onChangeText={[Function]}
            rejectResponderTermination={true}
            style={
              Object {
                "alignSelf": "center",
                "borderBottomColor": "gray",
                "borderBottomWidth": 1,
                "color": "grey",
                "fontSize": 18,
                "fontWeight": "bold",
                "marginLeft": 20,
                "width": 100,
              }
            }
            underlineColorAndroid="transparent"
            value="Nike"
          />
          <Text
            style={
              Object {
                "alignSelf": "center",
                "borderBottomColor": "gray",
                "color": "grey",
                "fontSize": 16,
                "fontWeight": "bold",
                "marginLeft": 10,
              }
            }
          >
            (9978.00%)
          </Text>
        </View>
        <Text
          style={
            Object {
              "color": "red",
              "fontSize": 10,
              "textAlign": "center",
            }
          }
        >
          
        </Text>
        <View
          style={
            Object {
              "color": "grey",
              "flexDirection": "row",
              "fontSize": 18,
              "fontWeight": "bold",
              "height": 35,
            }
          }
        >
          <Text
            style={
              Object {
                "alignSelf": "center",
                "color": "black",
                "fontSize": 18,
                "fontWeight": "bold",
                "marginTop": 15,
              }
            }
          >
            Full Condition Price:
          </Text>
          <TextInput
            allowFontScaling={true}
            clearButtonMode="while-editing"
            defaultValue="50"
            editable={false}
            maxLength={20}
            name="fullConditionPrice"
            rejectResponderTermination={true}
            style={
              Object {
                "alignSelf": "center",
                "color": "grey",
                "fontSize": 18,
                "fontWeight": "bold",
                "marginLeft": 10,
                "marginTop": 15,
              }
            }
            underlineColorAndroid="transparent"
            value="50£"
          />
        </View>
        <View
          style={
            Object {
              "color": "grey",
              "flexDirection": "row",
              "fontSize": 18,
              "fontWeight": "bold",
              "height": 35,
            }
          }
        >
          <Text
            style={
              Object {
                "alignSelf": "center",
                "color": "black",
                "fontSize": 18,
                "fontWeight": "bold",
                "marginTop": 15,
              }
            }
          >
            Resale Price:
          </Text>
          <TextInput
            allowFontScaling={true}
            clearButtonMode="while-editing"
            defaultValue="25"
            editable={false}
            maxLength={20}
            name="resalePrice"
            rejectResponderTermination={true}
            style={
              Object {
                "alignSelf": "center",
                "color": "grey",
                "fontSize": 18,
                "fontWeight": "bold",
                "marginLeft": 10,
                "marginTop": 15,
              }
            }
            underlineColorAndroid="transparent"
            value="25£"
          />
        </View>
      </View>
    </View>
    <View
      accessible={true}
      focusable={true}
      isTVSelectable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "alignSelf": "center",
          "backgroundColor": "#318F95",
          "borderRadius": 10,
          "height": 50,
          "justifyContent": "center",
          "marginTop": 20,
          "opacity": 1,
          "width": 190,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "white",
            "fontSize": 20,
            "fontWeight": "bold",
          }
        }
      >
        Register Item
      </Text>
    </View>
  </View>
  <Image
    resizeMode="stretch"
    source={1}
    style={
      Object {
        "height": "12%",
        "marginTop": "auto",
      }
    }
  />
</View>
`;
